File: 542-01-matrix/542-01-matrix.scala
import scala.collection.mutable._

object Solution {
    def updateMatrix(mat: Array[Array[Int]]): Array[Array[Int]] = {
        val moves = Seq((0, 1), (1, 0), (0, -1), (-1, 0))
        val r = mat.length
        val c = mat(0).length
        val res: Array[Array[Int]] = Array.fill[Array[Int]](r) { Array.fill[Int](c) {-1} }
        val q: Queue[(Int, Int, Int)] = Queue()
        
        for(i <- 0 to r - 1) {
            for(j <- 0 to c - 1) {
                if (mat(i)(j) == 0) {
                    q.enqueue((i, j, 0))
                }
            }
        }
        
        while(!q.isEmpty) {
            val p = q.dequeue
            if (res(p._1)(p._2) == -1) {
                res(p._1)(p._2) = p._3
                
                for(n <- moves) {
                    val nx = n._1 + p._1
                    val ny = n._2 + p._2
                    
                    if (nx >= 0 && ny >= 0 && nx < r && ny < c && res(nx)(ny) == -1)  {
                        q.enqueue((nx, ny, p._3 + 1))
                    }
                }
                
            }
        }
        
        res
    }
}

File: 542-01-matrix/542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int r = mat.length;
        int c = mat[0].length;
        
        int[][] res = new int[r][c];
        
        Queue<int[]> queue = new LinkedList<>();
        
        for(int i = 0; i < r; i++) {
            for(int j = 0; j < c; j++) {
                if (mat[i][j] == 0)  queue.offer(new int[] {i, j});
            }
        }
        
        int[][] dirs = new int[][] {
            {0, 1}, {0, -1}, {1, 0}, {-1, 0}
        };
        
        while(!queue.isEmpty()) {
            int[] loc = queue.poll();
            int v = res[loc[0]][loc[1]];
            
            for(int[] dir : dirs) {
                int x = loc[0] + dir[0];
                int y = loc[1] + dir[1];
                
                if (x >= 0 && x < r && y >= 0 && y < c && mat[x][y] == 1 && res[x][y] == 0) {
                    queue.offer(new int[] {x, y});
                    res[x][y] = v + 1;
                }
            }
        }
        
        return res;
    }
}